File: Runtime/src/Core/CoreScript/BCL/TaskCompletionSource.js
///////////////////////////////////////////////////////////////////////////////
// TaskCompletionSource

ss.TaskCompletionSource = function#? DEBUG TaskCompletionSource$##() {
	this.task = new ss.Task();
};
ss.TaskCompletionSource.prototype = {
	setCanceled: function#? DEBUG TaskCompletionSource$setCanceled##() {
		if (!this.task._cancel())
			throw 'Task was already completed.';
	},
	setResult: function#? DEBUG TaskCompletionSource$setResult##(result) {
		if (!this.task._complete(result))
			throw 'Task was already completed.';
	},
	setException: function#? DEBUG TaskCompletionSource$setException##(exception) {
		if (!this.task._fail(exception))
			throw 'Task was already completed.';
	},
	trySetCanceled: function#? DEBUG TaskCompletionSource$trySetCanceled##() {
		return this.task._cancel();
	},
	trySetResult: function#? DEBUG TaskCompletionSource$setResult##(result) {
		return this.task._complete(result);
	},
	trySetException: function#? DEBUG TaskCompletionSource$setException##(exception) {
		return this.task._fail(exception);
	}
};

ss.TaskCompletionSource.registerClass('ss.TaskCompletionSource');

File: Runtime/src/Libraries/CoreLib/Threading/Tasks/TaskCompletionSource.cs
﻿using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;
using System.Text;

namespace System.Threading.Tasks {
	[Imported(IsRealType = true)]
	[IgnoreGenericArguments]
	[ScriptNamespace("ss")]
	public class TaskCompletionSource<TResult> {
		public TaskCompletionSource() {
		}

		[IntrinsicProperty]
		public Task<TResult> Task { get { return null; } }

		public void SetCanceled() {
		}

		[NonScriptable]
		public void SetException(IEnumerable<Exception> exceptions) {
		}

		public void SetException(Exception exception) {
		}

		public void SetResult(TResult result) {
		}

		public bool TrySetCanceled() {
			return false;
		}

		[NonScriptable]
		public bool TrySetException(IEnumerable<Exception> exceptions) {
			return false;
		}

		public bool TrySetException(Exception exception) {
			return false;
		}

		public bool TrySetResult(TResult result) {
			return false;
		}
	}
}
